plugins/meta/vrf/main.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2020 CNI authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/vishvananda/netlink"
+
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/containernetworking/cni/pkg/types/current"
+	"github.com/containernetworking/cni/pkg/version"
+
+	"github.com/containernetworking/plugins/pkg/ns"
+	bv "github.com/containernetworking/plugins/pkg/utils/buildversion"
+)
+
+// VRFNetConf represents the vrf configuration.
+type VRFNetConf struct {
+	types.NetConf
+
+	// VRFName is the name of the vrf to add the interface to.
+	VRFName string `json:"vrfname"`
+}
+
+func main() {
+	skel.PluginMain(cmdAdd, cmdCheck, cmdDel, version.PluginSupports("0.4.0"), bv.BuildString("vrf"))
+}
+
+func cmdAdd(args *skel.CmdArgs) error {
+	conf, result, err := parseConf(args.StdinData)
+	if err != nil {
+		return err
+	}
+
+	if conf.PrevResult == nil {
+		return fmt.Errorf("missing prevResult from earlier plugin")
+	}
+
+	err = ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
+		vrf, err := findVRF(conf.VRFName)
+
+		if _, ok := err.(netlink.LinkNotFoundError); ok {
+			vrf, err = createVRF(conf.VRFName)
+		}
+
+		if err != nil {
+			return err
+		}
+
+		err = addInterface(vrf, args.IfName)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("cmdAdd failed: %v", err)
+	}
+
+	if result == nil {
+		result = &current.Result{}
+	}
+
+	return types.PrintResult(result, conf.CNIVersion)
+}
+
+func cmdDel(args *skel.CmdArgs) error {
+	conf, _, err := parseConf(args.StdinData)
+	if err != nil {
+		return err
+	}
+	err = ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
+		vrf, err := findVRF(conf.VRFName)
+		if err != nil {
+			return err
```

Comment:
@dcbw is not the DEL idempotent? should it return nil ?

Reply:
Ah, maybe only if err is a `LinkNotFoundError`?

Reply:
you need a more CNI guy here, but the spec says

> Plugins should generally complete a DEL action without error even if some resources are missing. For example, an IPAM plugin should generally release an IP allocation and return success even if the container network namespace no longer exists, unless that network namespace is critical for IPAM management. While DHCP may usually send a 'release' message on the container network interface, since DHCP leases have a lifetime this release action would not be considered critical and no error should be returned. For another example, the bridge plugin should delegate the DEL action to the IPAM plugin and clean up its own resources (if present) even if the container network namespace and/or container network interface no longer exist.

https://github.com/containernetworking/cni/blob/master/SPEC.md

Reply:
Thanks for the explanation!
Yes, I will modify the behavior so the DEL fails only if there's a real failure and not because the resources are not there.